drafts/str4d-overwinter-sighash/draft1.rst

```diff
@@ -0,0 +1,279 @@
+===========
+
+The key words "MUST" and "MUST NOT" in this document are to be interpreted as described in RFC 2119.
+
+
+Abstract
+========
+
+This proposal defines a new transaction digest algorithm for signature verification from the Overwinter
+network upgrade, in order to minimize redundant data hashing in verification, and to cover the input value by
+the signature.
+
+
+Motivation
+==========
+
+There are 4 ECDSA signature verification codes in the original Zcash script system: ``CHECKSIG``,
+``CHECKSIGVERIFY``, ``CHECKMULTISIG``, ``CHECKMULTISIGVERIFY`` ("sigops"). According to the sighash type
+(``ALL``, ``NONE``, or ``SINGLE``, possibly modified by ``ANYONECANPAY``), a transaction digest is generated
+with a double SHA256 of a serialized subset of the transaction, and the signature is verified against this
+digest with a given public key. The detailed procedure is described in a Bitcoin Wiki article. [#wiki-checksig]_
+
+Unfortunately, there are at least 2 weaknesses in the original SignatureHash transaction digest algorithm:
+
+* For the verification of each signature, the amount of data hashing is proportional to the size of the
+  transaction. Therefore, data hashing grows in O(n\ :sup:`2`) as the number of sigops in a transaction
+  increases. While a 1 MB block would normally take 2 seconds to verify with an average computer in 2015, a
+  1MB transaction with 5569 sigops may take 25 seconds to verify. This could be fixed by optimizing the digest
+  algorithm by introducing some reusable “midstate”, so the time complexity becomes O(n). [#quadratic]_
+
+* The algorithm does not involve the value being spent by the input. This is usually not a problem for online
+  network nodes as they could request for the specified transaction to acquire the output value. For an
+  offline transaction signing device ("cold wallet"), however, the lack of knowledge of input amount makes it
+  impossible to calculate the exact amount being spent and the transaction fee. To cope with this problem a
+  cold wallet must also acquire the full transaction being spent, which could be a big obstacle in the
+  implementation of a lightweight, air-gapped wallet. By including the input value of part of the transaction
+  digest, a cold wallet may safely sign a transaction by learning the value from an untrusted source. In the
+  case that a wrong value is provided and signed, the signature would be invalid and no funds would be lost.
+  [#offline-wallets]_
+
+Deploying the aforementioned fixes in the original script system is not a simple task.
+
+
+Specification
+=============
+
+A new transaction digest algorithm is defined, but only applicable from the Overwinter upgrade block height
+[#ZIP0000]_::
+  [TODO: Pick one] BLAKE2[b-256|s] of the serialization of:
+    1. nVersion of the transaction (4-byte little endian)
+    2. hashPrevouts (32-byte hash)
+    3. hashSequence (32-byte hash)
+    4. hashOutputs (32-byte hash)
+    5. hashJoinSplits (32-byte hash)
```

Comment:
@daira I was thinking that we might end up having separate hashes for input and output Sapling circuits, ie:

```
5. hashJoinSplits (32-byte hash)
6. hashShieldedInputs (32-byte hash)
7. hashShieldedOutputs (32-byte hash)
```

I'm not sure whether we need this kind of flexibility though, in the way that it is useful for the different transparent sighash types (because I expect we'd want to implement those in a zero-knowledge way, which doing inside a visible signature would not be).

Note that we will need to have another ZIP with a newer SigHash specification anyway when we address transaction malleability.

Reply:
We discussed this in a pairing, and decided that we don't need to add anything for Sapling in advance; we can conditionally add it to the algorithm (gated on transaction version) when we add Sapling.